### ISSUE TITLE
Need for subscription on linked maps

### ISSUE THREAD

**Issue by jacogr** [polkadot-js/api#802]
(Marking as question since it is not really confirmed and may just be due to me not quite grasping some basics)

Let's start at the top, I believe this is needed: https://github.com/polkadot-js/api/blob/master/packages/api/src/Base.ts#L554 - as new entries get added, the head is updated, so it nicely triggers the reload of all/

However, not convinced about this - https://github.com/polkadot-js/api/blob/master/packages/api/src/Base.ts#L505

The problem is that as the entry changes (say an entry was removed in the middle), I'm not sure how this actually caters for refreshing that part of the list and/or the whole list. Since at on subsequent loads we are basically done, I am not sure how the list gets refreshed. 

TL;DR - Can see how adding a new entry causes a refresh of the list (and can visibly see this with the apps UI), however for stuff that gets removed from the list, cannot quite wrap around the refresh thereof.

This could be related to https://github.com/polkadot-js/apps/issues/857 (however that could very well be a UI bug, things are still WIP and not 100% clean there)

cc @xlc @amaurymartiny  - you may be able to help guide me along the right track. This _looks_ correct, but not sure how/if it works in practice since with these refreshes can be problematic

**Comment by xlc** [comment]
I think the only way to test if it works is have a testing module so we can see how the current code handles all the scenarios.

**Comment by jacogr** [comment]
100% - in this case I just (also) cannot wrap my head around what happens internally (between the subs and RxJs and the conversions at the end) when the previous/next changes.

**Comment by xlc** [comment]
Maybe should provide some diagnostic function to print all the subscriptions to help understanding what is happening.
I am expecting people are going to have leaky subscription issue.

**Comment by jacogr** [comment]
Ok following @xlc with some debugging, there are some gaps, but not surprising since this is very messy. Tracking nominators and entering, removing.

Initial -
```
console.log packages/api/src/Base.ts:604
    head changed

  console.log packages/api/src/Base.ts:554
    list changed: 5GGxQvLetFxUFC3K8rtQxXcVdnMN7gyAkGaMLxaS8Ra1eQNn

  console.log packages/api/src/Base.ts:554
    list changed: 5CqWLoyv1joRsEWEqY1U2UCEHho2uwMGfde3wgGXGEi2ZNtz

  console.log packages/api/src/Base.ts:554
    list changed: 5G9YqP3mR96h7r88DD4n2MVSgG5z4sHD5ooiuZ4cYdqWDF13

  console.log packages/api/src/Base.ts:554
    list changed: 5FHgmZoZUttLjvWCXuSWnwH2pNNB56CxLipfT3wyCyLcwLBM

  console.log packages/api/test/e2e/promise-alex.spec.js:42
    *** 1 nominators() [ [ '5GGxQvLetFxUFC3K8rtQxXcVdnMN7gyAkGaMLxaS8Ra1eQNn',
        '5CqWLoyv1joRsEWEqY1U2UCEHho2uwMGfde3wgGXGEi2ZNtz',
        '5G9YqP3mR96h7r88DD4n2MVSgG5z4sHD5ooiuZ4cYdqWDF13',
        '5FHgmZoZUttLjvWCXuSWnwH2pNNB56CxLipfT3wyCyLcwLBM' ],
      [ [ '5G1EFrPw1JQtzZzRZmLV4gsrYahaxo2CbdyAXLzgmDxsgJYS' ],
        [ '5F4PGNWujpW1572z6tqeida4tEB8gM7THqpmFKpkBTYLohsX' ],
        [ '5FXiEsPgPMzKk6LPJFMPhfBT5UQzPKKVRWMnbzGu5GsE4G5T' ],
        [ '5DPW1n4q2THUjKGj3QKkcqdh6oxY1bmLsSQ7t8FuiuNtRoUc' ] ] ]
```

Nonimating -

```
  console.log packages/api/src/Base.ts:554
    list changed: 5GGxQvLetFxUFC3K8rtQxXcVdnMN7gyAkGaMLxaS8Ra1eQNn

  console.log packages/api/src/Base.ts:604
    head changed

  console.log packages/api/src/Base.ts:554
    list changed: 5D1yhWoGPmrFy5AKcS5BqrarjGUjiUGyb2BzwUpxrgZxz6BV

  console.log packages/api/src/Base.ts:554
    list changed: 5GGxQvLetFxUFC3K8rtQxXcVdnMN7gyAkGaMLxaS8Ra1eQNn

  console.log packages/api/src/Base.ts:554
    list changed: 5CqWLoyv1joRsEWEqY1U2UCEHho2uwMGfde3wgGXGEi2ZNtz

  console.log packages/api/src/Base.ts:554
    list changed: 5G9YqP3mR96h7r88DD4n2MVSgG5z4sHD5ooiuZ4cYdqWDF13

  console.log packages/api/src/Base.ts:554
    list changed: 5FHgmZoZUttLjvWCXuSWnwH2pNNB56CxLipfT3wyCyLcwLBM

  console.log packages/api/test/e2e/promise-alex.spec.js:42
    *** 2 nominators() [ [ '5GGxQvLetFxUFC3K8rtQxXcVdnMN7gyAkGaMLxaS8Ra1eQNn' ],
      [ [ '5G1EFrPw1JQtzZzRZmLV4gsrYahaxo2CbdyAXLzgmDxsgJYS' ] ] ]
```

Unnominating -

```
  console.log packages/api/src/Base.ts:554
    list changed: 5GGxQvLetFxUFC3K8rtQxXcVdnMN7gyAkGaMLxaS8Ra1eQNn

  console.log packages/api/src/Base.ts:604
    head changed

  console.log packages/api/src/Base.ts:554
    list changed: 5GGxQvLetFxUFC3K8rtQxXcVdnMN7gyAkGaMLxaS8Ra1eQNn

  console.log packages/api/src/Base.ts:554
    list changed: 5CqWLoyv1joRsEWEqY1U2UCEHho2uwMGfde3wgGXGEi2ZNtz

  console.log packages/api/src/Base.ts:554
    list changed: 5G9YqP3mR96h7r88DD4n2MVSgG5z4sHD5ooiuZ4cYdqWDF13

  console.log packages/api/src/Base.ts:554
    list changed: 5FHgmZoZUttLjvWCXuSWnwH2pNNB56CxLipfT3wyCyLcwLBM

  console.log packages/api/test/e2e/promise-alex.spec.js:42
    *** 3 nominators() [ [ '5GGxQvLetFxUFC3K8rtQxXcVdnMN7gyAkGaMLxaS8Ra1eQNn' ],
      [ [ '5G1EFrPw1JQtzZzRZmLV4gsrYahaxo2CbdyAXLzgmDxsgJYS' ] ] ]
```

So basically the final results don't take all the entries into account.
